Partitioning refactor

dafyddstephenson opened the pull request:
This PR makes some internal changes to partitioning:

- introduces a `ROMSPartitioning` class 
- replaces `ROMSInputDataset.partitioned_files` with `ROMSInputDataset.partitioning` (an instance of the above class) to track locally partitioned netCDF files more effectively than just listing them
- defines several utility functions to modularise methods that fetch and partition datasets.
- Adds (`Optional[int]`) `source_np_xi` and `source_np_eta` parameters to `ROMSInputDataset.__init__` to specify that the source of a dataset is already partitioned, allowing users to supply pre-partitioned data to instantiate `ROMSInputDataset`s (#247)
- Adds checks that allow C-Star to skip unnecessary repeat calls to `pre_run` or `partition` (#214)
- Catch use cases like attempting to partition data multiple times with different parameters 

- [x] Closes #247 
- [x] Closes #214 
- [x] Tests added
- [x] Tests passing
- [x] Full type hint coverage
- [ ] Changes are documented in `docs/releases.rst`
- [ ] New functions/methods are listed in `api.rst`
- [ ] New functionality has documentation

Commit:
separate out yaml logic from ROMSInputDataset.get() into private _get_from_yaml method

Commit:
separate out download/symlink logic in InputDataset.get and private ROMSInputDataset._get_from_partitioned_source method

Commit:
Add tracking of partitioned files, prevent repeat calls to partition, update unit tests

Commit:
fix bug re. repeated pre_run() calls

Commit:
- Add Partitioning class and ROMSInputDataset.partitioning attribute
- More stringent checking of whether ROMSInputDataset.partition should be repeated
- Add check for whether all ROMSInputDataset.partitioning entries match ROMSSimulation CPU distribution

Commit:
update unit tests

Commit:
add source partitioning to blueprint

Commit:
add overwrite_existing_files option to ROMSSimulation.pre_run

Commit:
update unit tests

Commit:
tidy up

Commit:
add unit tests to re-establish 100% coverage of roms/input_dataset.py

Commit:
remove unused mock

ScottEilerman commented on cstar/base/input_dataset.py:
```diff
@@ -261,39 +261,60 @@ def get(self, local_dir: str | Path) -> None:
+        )
+
+        self.working_path = target_path
+        self._local_file_hash_cache.update({target_path: computed_file_hash})  # 27
```
`# 27` 🤔

dafyddstephenson replied:
IYKYK

ScottEilerman commented on cstar/roms/input_dataset.py:
```diff
@@ -27,24 +70,49 @@ class ROMSInputDataset(InputDataset, ABC):
+        self.partitioning: Optional[ROMSPartitioning] = None
+
+    @property
+    def source_partitioning(self):
```
```suggestion
    def source_partitioning(self) -> tuple[int,int] | None:
```

ScottEilerman commented on cstar/roms/input_dataset.py:
```diff
@@ -53,7 +121,9 @@ def __repr__(self) -> str:
 
-    def partition(self, np_xi: int, np_eta: int):
+    def partition(
+        self, np_xi: int, np_eta: int, overwrite_existing_files: bool = False
```
I was going to add a comment that, if `overwrite_existing_files is True`, we should maybe clear out the old files just to make sure we don't end up with a confusing state where some are overwritten but some others may be left over from the previous partitioning. But, now I'm wondering if we should just not support overwriting here at all, and make people manually clean up the old files if they're going to do so? Thoughts?

dafyddstephenson replied:
I considered this, though it is just a tidying thing: if, e.g. we first partition into 25 tiles and then 15, only the first 15 will be used, so 16-25 will just be hanging around doing nothing, in an ugly but ultimately harmless way. I also considered moving the existing files to a temporary location and having a `finally` to return them in case of failure, which covers both cases. WDYT?

ScottEilerman replied:
Yeah, I just imagine it will be easily forgotten and someone will look at the directly later and see 25 files, and not know that 16-25 are now totally incompatible with 1-15.

I guess the question to me is: how often is someone going to hit this "I want to intentionally change partitioning" condition? If it's often enough that we should handle it for them, we should handle it well. My assumption (which could be wrong) is that it will be rare enough that we can just throw an error for now and tell them to clean out the old files themselves.

dafyddstephenson replied:
Agree, I just took the "well they shouldn't be peeking at the files, C-Star is taking care of it for them" line. But either way it's a bit gross and not completely insane to do the temporary move thing, so I'll just add that.

dafyddstephenson replied:
[8ed114f](https://github.com/CWorthy-ocean/C-Star/pull/252/commits/8ed114f075c136225a9ff46911fe466f7dccad4f)

ScottEilerman commented on cstar/roms/input_dataset.py:
```diff
@@ -158,10 +235,78 @@ def get(
+        self, local_dir: Path, source_np_xi: int, source_np_eta: int
+    ) -> None:
+        n_source_partitions = source_np_xi * source_np_eta
+        ndigits = len(str(n_source_partitions))
```
This is genius. I would have done something over-complicated with division and `mod`.

ScottEilerman commented on cstar/base/input_dataset.py:
```diff
@@ -261,39 +261,60 @@ def get(self, local_dir: str | Path) -> None:
+        self._local_file_stat_cache.update({target_path: target_path.stat()})
+
+    @staticmethod
+    def _symlink_or_download_from_source(
```
Unless there's some usage I'm not seeing, it seems like we can drop `@staticmethod`, since we only ever call this from within the class and are often passing it `self` variables for location_type, logger, etc.

ScottEilerman replied:
After looking a little more, I see that in the two calls, some of the variables get `self.something` in one but a different value in the other, so it makes sense that they need to be passed in. It looks like the two I mentioned above are always the same `self` value though and could be dropped.

ScottEilerman replied:
This could wait until a later refactor if we want. In the next blueprint iteration, we will probably split the symlink and download functionality into two subclasses anyway (along the `if location_type` branches you have in here already).

ScottEilerman reviewed (commented):
A few cleanup comments, I think the only big decision is whether we bother supporting the overwrite feature (and if so, whether we need to add some safeguards).

dafyddstephenson committed:
typehint source_partitioning property

Co-authored-by: ScottEilerman <scott.eilerman@cworthy.org>

Commit:
Tidying, improvements and bugfixes
- Modularise 'ROMSInputDataset.partition'
- Add strategy to backup existing partitioned files and restore if attempt to create new files fails
- Change init approach for pre-pratitioned ROMSInputDataset - location should be /first partitioned file/

Commit:
fix unit tests

Commit:
restore 100% unit test coverage

Commit:
fix pre-commit issue

ScottEilerman commented on cstar/roms/input_dataset.py:
```diff
@@ -136,73 +137,134 @@ def partition(
+                ]
+            return id_files_to_partition
+
+        def partition_files(files: list[Path]) -> list[Path]:
```
The files arg here isn't used. It still ends up working because `id_files_to_partition` ends up in scope when it's called. I usually prefer private functions at the same level over inner functions for this reason, because the scope can get confusing. Just a preference though, there's reasons to prefer keeping all of the logic encapsulated as well, if the private functions don't get used anywhere else.

dafyddstephenson replied:
Good catch - fixed ([bfdbe45](https://github.com/CWorthy-ocean/C-Star/pull/252/commits/bfdbe45b4fc2bbed0cadc1f9de13747729754150)). The main reason for nesting these ones was just organisation/neatness (too much logic to sprawl it all out at once, too exclusive to the parent function to define at the class level). There is one example that's re-used, so is class-level: `_update_partitioning_attribute`.

ScottEilerman commented on cstar/roms/input_dataset.py:
```diff
@@ -230,6 +292,8 @@ def get(
         else:
             working_path_parent = self.working_path.parent
+        # import pdb
+        # pdb.set_trace()
```
Missed some cleanup here.

dafyddstephenson replied:
No idea how this got in here, wasn't me. I always tidy up comments and never have to debug, perfect code flows directly from my mind into the keys. Took it out ([c036584](https://github.com/CWorthy-ocean/C-Star/pull/252/commits/c0365840387410378552824388e5f1e0b65c0d0c)) but we should definitely find who's responsible 🥷

Commit:
fix typo in partition_files fn

Commit:
tidy up stray comments
